File: backend/app/api/part_numbers.py
"""PartNumber master endpoints (Phase 4 — Work Order intake, Add Part).

HTTP surface for the PN lookup/create step of the Add Part flow
(GUI_DESIGN §11.2/§11.3) and the minimal barcode-label capability of
Phase 4: enough data to view/print the derived ``PF:PN:<pn>`` label.

Routes stay thin orchestration: request schemas validate shape only
(``extra="forbid"``), the Application layer owns normalization,
create-on-first-use, the audit protocol and the transaction, and the
central handlers in ``app.api.errors`` translate typed failures.

Deliberate surface decisions:

- ``barcode_value`` appears only in responses — the PN barcode is
  fully derived from the canonical PN and never stored or entered.
- Lookup uses query parameters (``search`` contains-match, ``number``
  exact canonical resolution) instead of a path segment: the PN is an
  opaque arbitrary string, so it never travels as a URL path value.
- ``POST /part-numbers`` is create-or-reuse: an existing canonical PN
  returns the existing master (200), first valid use creates it (201)
  with its ``CREATED`` audit row in the same transaction. There is no
  DELETE and no update — master metadata management is Phase 13.
"""

import datetime

from fastapi import APIRouter, Response
from pydantic import BaseModel, ConfigDict

from app.api.dependencies import SessionDep
from app.application import part_numbers
from app.infrastructure.models import PartNumber

router = APIRouter(prefix="/api")


class PartNumberResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    # The canonical uppercase PN — the identity and natural key.
    part_number: str
    # Derived label data: PF:PN:<canonical-part-number>.
    barcode_value: str
    created_at: datetime.datetime
    updated_at: datetime.datetime


class PartNumberCreateRequest(BaseModel):
    """Only the PN itself — the audit ``actor_reference`` is never
    client-writable: it stays NULL from this HTTP surface until an
    authenticated identity exists (Phase 14)."""

    model_config = ConfigDict(extra="forbid")

    part_number: str


@router.get("/part-numbers")
def list_part_numbers(
    session: SessionDep, search: str | None = None, number: str | None = None
) -> list[PartNumberResponse]:
    """PN lookup: ``number`` resolves one exact canonical PN (empty list
    on a miss — the Add Part flow then offers creation), ``search``
    filters by contains-match.

    A ``search`` (or unfiltered) listing is bounded server-side at
    ``part_numbers.SEARCH_RESULT_LIMIT`` masters; an exact ``number``
    resolution is never bounded away, so a short but valid canonical PN
    still resolves.
    """
    matches: list[PartNumber]
    if number is not None:
        master = part_numbers.resolve_part_number(session, number)
        matches = [master] if master is not None else []
    else:
        matches = part_numbers.list_part_numbers(session, search=search)
    return [PartNumberResponse.model_validate(master) for master in matches]


@router.post("/part-numbers")
def create_part_number(
    body: PartNumberCreateRequest, session: SessionDep, response: Response
) -> PartNumberResponse:
    master, created = part_numbers.create_part_number(session, body.part_number)
    response.status_code = 201 if created else 200
    return PartNumberResponse.model_validate(master)

File: frontend/src/api/part-numbers.ts
// PartNumber master API (Phase 4 — Add Part lookup, GUI_DESIGN §11.2/§11.3).
//
// The PN string itself is the identity: lookup travels as query
// parameters (never a URL path segment), `number` resolves one exact
// canonical PN and `search` is a contains-match. The barcode value is
// fully derived (`PF:PN:<canonical-PN>`) and only ever read.
//
// Creation is owned by the Work Order save transaction
// (create-on-first-valid-use) — the Add Part flow only needs lookup
// plus the client-side mirror of the canonical normalization
// (`normalizePartNumber`, `views/scan-station/barcode.ts`); server
// validation remains the authority.
//
// Production-safe: no mock data, no framework imports.

import { apiRequest } from './client';

export interface PartNumberMaster {
  /** The canonical uppercase PN — the identity and natural key. */
  partNumber: string;
  /** Derived label data: `PF:PN:<canonical-part-number>`. */
  barcodeValue: string;
}

interface PartNumberWire {
  part_number: string;
  barcode_value: string;
}

function toMaster(wire: PartNumberWire): PartNumberMaster {
  return { partNumber: wire.part_number, barcodeValue: wire.barcode_value };
}

/** Contains-search over the PN masters (all of them for a blank query). */
export async function searchPartNumbers(
  search: string,
): Promise<PartNumberMaster[]> {
  const query = search.trim()
    ? `?search=${encodeURIComponent(search.trim())}`
    : '';
  const wires = await apiRequest<PartNumberWire[]>(`/api/part-numbers${query}`);
  return wires.map(toMaster);
}

/**
 * Exact canonical resolution of one PN, or null when no master exists
 * yet — the Add Part flow then offers explicit creation on first use.
 */
export async function resolvePartNumber(
  partNumber: string,
): Promise<PartNumberMaster | null> {
  const wires = await apiRequest<PartNumberWire[]>(
    `/api/part-numbers?number=${encodeURIComponent(partNumber)}`,
  );
  return wires.length > 0 ? toMaster(wires[0]) : null;
}
